File: src/samples/Counters/Views/MouseWheelGesture.cs
using System;
using System.Windows.Input;

namespace Views
{
    public class MouseWheelGesture : MouseGesture
    {
        protected MouseWheelGesture(Func<MouseWheelEventArgs, bool> predicate,
            ModifierKeys modifiers = ModifierKeys.None) : base(MouseAction.WheelClick, modifiers)
        {
            _predicate = predicate;
        }

        private readonly Func<MouseWheelEventArgs, bool> _predicate;

        public override bool Matches(object targetElement, InputEventArgs inputEventArgs)
        {
            if (!base.Matches(targetElement, inputEventArgs)) return false;
            var args = inputEventArgs as MouseWheelEventArgs;
            return args != null && _predicate(args);
        }
    }

    public class MouseWheelUp : MouseWheelGesture
    {
        public MouseWheelUp(ModifierKeys modifiers = ModifierKeys.None) : base(e => e.Delta > 0, modifiers)
        {
        }

        public MouseWheelUp() : this(ModifierKeys.None) { }
    }

    public class MouseWheelDown : MouseWheelGesture
    {
        public MouseWheelDown(ModifierKeys modifiers = ModifierKeys.None) : base(e => e.Delta < 0, modifiers)
        {
        }

        public MouseWheelDown() : this(ModifierKeys.None) { }
    }
}

File: samples/Counters/Helpers/MouseWheelGesture.fs
namespace Helpers.Gestures

open System
open System.Windows.Input

type MouseWheelGesture (predicate ,modifiers: ModifierKeys) =
    inherit MouseGesture(MouseAction.WheelClick, modifiers)
    member private this.Predicate = predicate
    override this.Matches (targetElement, inputEventArgs) =
        if (not (base.Matches(targetElement, inputEventArgs))) then
            false
        else
            let args = inputEventArgs :?> MouseWheelEventArgs
            match args with
            | null -> false
            | _ -> (this.Predicate args)

module methods =
    let greaterThan limit (e:MouseWheelEventArgs) = e.Delta > limit
    let lessThan limit (e:MouseWheelEventArgs) = e.Delta < limit

type MouseWheelUp (modifiers) =
    inherit MouseWheelGesture((methods.greaterThan 0), modifiers)
    new() = 
        MouseWheelUp(ModifierKeys.None)

type MouseWheelDown (modifiers) =
    inherit MouseWheelGesture((methods.lessThan 0), modifiers)
    new() = 
        MouseWheelDown(ModifierKeys.None)
